cloudfoundry/cf-java-client
# Upgrade JUnit4 to JUnit5

In order to complete this upgrade, I followed public documentation, here:

    https://docs.openrewrite.org/running-recipes/popular-recipe-guides/migrate-from-junit-4-to-junit-5
    
I employed OpenRewrite to perform initial bulk operation with:

```
mvn -U org.openrewrite.maven:rewrite-maven-plugin:run \
  -Drewrite.recipeArtifactCoordinates=org.openrewrite.recipe:rewrite-spring:5.1.7 \
  -Drewrite.activeRecipes=org.openrewrite.java.spring.boot2.SpringBoot2JUnit4to5Migration
```

I also upgraded dependencies and plugins in root `pom.xml`:

* Jackson to 2.16.1
* JUnit Jupiter to 5.10.1
* Maven Compiler to 3.12.1
* Maven Surefire to 3.2.3

Afterwards:

* I adjusted imports on `AbstractRestTest.java`, then replaced a reference to `org.junit.ComparisonFailure` (no longer available in JUnit5) which was itself is a subclass of `java.lang.AssertionError`.
* I had to update `DefaultApplicationsTest.java`, `DefaultSpacesTest.java` where `Supplier<T>` anonymous impl for `get()` method did not have correct visibility.

Result:

* Tests pass for Java 8 and Java 11
  * Maven profile is activated for a installed JDK version detected
  * Defaults to Spring Boot to 2.7.17
  * See https://github.com/pacphi/cf-java-client/actions/runs/7371626056  
* Tests fail for Java 17 and 21
  * Maven profile is activated for a installed JDK version detected
  * Sets version of Spring Boot dependencies to 3.2.1
  * See https://github.com/pacphi/cf-java-client/actions/runs/7371626055
  * Observe the following issue targeting Java 17+, here https://github.com/immutables/immutables/issues/1339
